### PR TITLE
feat: [MR-579] Enable reject signals for requests

### PR DESCRIPTION
**Pull request by stiegerc**
Enables reject signals for requests for `CertificationVersion::V19` and above; keeps generating reject responses for versions below that.

Adds new tests as modification of the old tests; old tests are kept around with a `legacy_` prefix. Adapts one state machine test such that canisters are stopped rather than just make them stop making new calls, this provokes reject signals for requests thereby giving us a high level check on this functionality while leaving the test intact otherwise.
Adds one new state machine test for a canister migration to check all calls are concluded for this scenario.

### CODE REVIEW

File: rs/messaging/src/routing/stream_handler.rs [hunk 1]
```diff
@@ -735,36 +731,20 @@ impl StreamHandlerImpl {
...
+                Some((reason, msg)) => {
+                    // Unable to induct a message, push reject signal.
+                    stream.push_reject_signal(reason);
+                    debug_assert!(matches!(
+                        (reason, msg),
+                        // All reasons are legal for requests.
+                        (_, RequestOrResponse::Request(_))
+                            // Only `CanisterMigrating` is legal for responses.
+                            | (RejectReason::CanisterMigrating, RequestOrResponse::Response(_))
+                    ));
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler.rs:747]: I put this into a `debug_assert!()` because this couldn't fail because of any incoming stream slices or anything like that, it's just the way `induct_message_impl()` is written. We should keep it, put having this fire during tests is good enough.

File: rs/messaging/src/routing/stream_handler.rs [hunk 2]
```diff
@@ -793,7 +773,7 @@ impl StreamHandlerImpl {
         msg_type: &str,
         state: &mut ReplicatedState,
         available_guaranteed_response_memory: &mut i64,
-    ) -> Result<(), (StateError, RequestOrResponse)> {
+    ) -> Option<(RejectReason, RequestOrResponse)> {
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler.rs:776]: Returning `Result` here was a bit awkward since it would return `Ok` for cases where it failed to induct but we didn't produce a reject response, i.e. it didn't mean 'we successfully inducted the message'. Essentially I had made it like this because then it was using the same return type `ReplicatedState::push_input()` does and we needed both the message and the `StateError` to generate the reject response.
But now it returns a reject reason instead, well basically a reject signal. So the function now may or may not return a reject signal, which says nothing about whether we failed or logged an error.

The idea was initially to have the only change be in this place for this PR by replacing generating a reject response with pushing a reject signal, but now it seems generating the signal inside `induct_message_impl()` makes more sense.

File: rs/replicated_state/src/replicated_state.rs [hunk 3]
```diff
@@ -99,12 +99,6 @@ pub enum StateError {
     /// Canister is stopping, only accepting responses.
     CanisterStopping(CanisterId),
 
-    /// Message enqueuing failed due to canister migration.
-    CanisterMigrating {
-        canister_id: CanisterId,
-        host_subnet: SubnetId,
-    },
```
Comment (stiegerc) [inline rs/replicated_state/src/replicated_state.rs:106]: I removed this variant again because with `induct_message_impl()` returning a reject signal instead, this is no longer needed. It made sense before this PR because it allowed to unify the way reject responses were generated, but this kind of reject response is now gone altogether. The only reject responses we still have are locally generated upon observing a reject signal. The only way those get into a stream is during canister migration where they can't be inducted on the former host subnet.

**Review by stiegerc — COMMENTED** [review]
Made some small drive-by refactors to the tests also. Overall it has also simplified the tests quite a bit.

We pass `available_guaranteed_response_memory` into most functions, which is supposed to track the changes in memory. However, it doesn't track gc'ing responses in streams, which I still find rather confusing.
But anyways, since locally generated reject responses are inducted directly and are never part of any streams (except during canister migration), this has actually indirectly improved the memory tracking as well. And by doing that it is also simplified the tests.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 4]
```diff
@@ -291,41 +279,27 @@ fn induct_loopback_stream_reroute_response() {
...
-                    - reject_response_count_bytes as i64,
-                available_guaranteed_response_memory,
-            );
-            // Not equal, because the computed available memory does not account for the
-            // reject response (since it's from a canister no longer hosted by the subnet).
-            assert!(
-                stream_handler.available_guaranteed_response_memory(&inducted_state)
-                    >= available_guaranteed_response_memory
+                available_guaranteed_response_memory + inducted_response_count_bytes as i64,
+                stream_handler.available_guaranteed_response_memory(&inducted_state),
```
Comment (stiegerc) [inline rs/messaging/src/routing/stream_handler/tests.rs:293]: To be honest, I find the way I wrote this here both snappier and less confusing. It shows directly that the computed value and the tracked value differ by a few bytes because garbage collection is not tracked properly. It also required less thinking on my side to get it right.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:293]: There are other places where I could rewrite it like this if people agree with my assessment.

**Comment by alin-at-dfinity** [comment]
> We pass `available_guaranteed_response_memory` into most functions, which is supposed to track the changes in memory. However, it doesn't track gc'ing responses in streams, which I still find rather confusing.

This is because we don't know whether a GC-ed message belonged to an existing canister or not (e.g. the canister could have been deleted or migrated). And because when we compute the message memory used / available we only look at current canisters, releasing memory belonging to a no-longer-existent canister would be incorrect (and might even technically underflow). Instead we're conservative and choose to err on the side of less available memory.

If we were fine with not charging canisters for the messages they have in streams, we could simplify the logic by (i) only looking at messages in canisters' queues when computing per-canister memory usage; and (ii) adding the size of the streams only to the subnet memory usage. But that would be a relatively involved change.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 4 above)
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler/tests.rs:293]: I suppose it's OK either way: as it was before, it was more verbose, but it actually looked for an absolute value (and it allowed the reader to track exactly how much memory was allocated for what); as it is now, it's more concise, but it's merely a relative comparison of the current actual usage vs. a pessimistic estimate (so you wouldn't find out from this assert alone if e.g. the implementation somehow forgot to account for the reject response).

Given the above, I would rather have a mix of absolute and relative comparisons than only relative comparisons everywhere.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 5]
```diff
@@ -2090,24 +1899,20 @@ fn induct_stream_slices_with_messages_to_migrating_canister() {
...
-                    reject_response,
                 ],
                 // ... a `signals_end` advanced by 2...
                 signals_end: 45,
-                // ...and a reject signal for the response @44.
-                reject_signals: vec![RejectSignal::new(
-                    RejectReason::CanisterMigrating,
-                    44.into(),
-                )],
+                // ...and a reject signals for the request @43 and the response @44.
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1905]:
```suggestion
                // ...and reject signals for the request @43 and the response @44.
```

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 6]
```diff
@@ -2117,18 +1922,11 @@ fn induct_stream_slices_with_messages_to_migrating_canister() {
             );
 
             assert_eq!(expected_state, inducted_state);
-
-            // One reject response enqueued.
             assert_eq!(
-                initial_available_guaranteed_response_memory - reject_response_count_bytes as i64,
+                stream_handler.available_guaranteed_response_memory(&inducted_state),
                 available_guaranteed_response_memory
             );
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler/tests.rs:1928]: E.g. In this particular case, I believe the earlier assert(s) were more useful than the current one.

Actually, how come we went from `assert!(a >= b)` (which, at least according to the comment was actually always `<`) to `assert_eq!(a, b)`?

Same below.

File: rs/messaging/src/routing/stream_handler.rs [hunk 7]
```diff
@@ -735,36 +731,20 @@ impl StreamHandlerImpl {
...
-                    *available_guaranteed_response_memory -=
-                        stream.push(generate_reject_response(&request, code, err.to_string()))
-                            as i64;
-                    stream.push_accept_signal()
-                }
-                Err((_, RequestOrResponse::Response(_))) => {
-                    unreachable!("No signals are generated for response induction failures except for CanisterMigrating");
+                Some((reason, msg)) => {
+                    // Unable to induct a message, push reject signal.
+                    stream.push_reject_signal(reason);
```
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:740]: You could have a certification version < 19 if we set `CURRENT_CERTIFICATION_VERSION` to an older version. This is entirely possible if we find an issue in `V19` (or `V18`, or whatever).

But more generally, for as long as we have a version defined in the `CertificationVersion` enum, we need to be able to produce a canonical state as defined by that version.

And yes, that may mean that you have to keep around old tests.

The way to get rid of all of this cruft is to remove older `CertificationVersions` (maybe keeping a couple of "fake" ones so we can retain the existing test fixtures).

File: rs/messaging/src/routing/stream_handler.rs (hunk 1 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:747]: It may be safe, but also pointless. The only way this can fail is if you change the structure in the code and this is caught in tests.

File: rs/messaging/src/routing/stream_handler/tests.rs [hunk 8]
```diff
@@ -1656,156 +1630,6 @@ fn check_stream_handler_locally_generated_reject_response_unknown() {
     );
 }
 
-// TODO: [MR-579] Remove this fixture and the tests that depend on it, since reject responses of
-// this type are no longer produced.
-/// Common implementation for tests checking reject responses generated by the `StreamHandler`
-/// directly.
-fn check_stream_handler_generated_reject_response_impl(
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:1663]: I put them back in and made them check for the reject signal, but I can't use `induct_loopback_stream()` for this directly, because that garbage collects the signals immediately. So I just induct the loopback stream as a slice instead.

For some signal types you could use `induct_loopback_stream()` because then we should see a reject response generated for the signal during the garbage collecting phase, but this is only true for some signals. For instance, with a canister migration, the reject response is routed into a stream instead and other constellations wouldn't even be possible at all such as `CanisterStopped` since there can't be a request in-flight from a canister that is stopped. What happens here is that the reject response is silently dropped.
You could also check for that but not in a unified way for all reject reasons and that's also not what these tests were supposed to test in the first place. Even before the tests were generating impossible constellations, except that the reject response ended up in the stream instead so the problem was circumvented.

File: rs/messaging/src/routing/stream_handler.rs [hunk 9]
```diff
@@ -824,7 +804,22 @@ impl StreamHandlerImpl {
...
+                                    }
+                                    StateError::CanisterStopped(_) => RejectReason::CanisterStopped,
+                                    StateError::CanisterStopping(_) => {
+                                        RejectReason::CanisterStopping
+                                    }
+                                    StateError::QueueFull { .. } => RejectReason::QueueFull,
+                                    StateError::OutOfMemory { .. } => RejectReason::OutOfMemory,
+                                    StateError::NonMatchingResponse { .. }
+                                    | StateError::BitcoinNonMatchingResponse { .. } => {
+                                        RejectReason::Unknown
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:819]: DIdn't we introduce `RejectReason::Unknown` precisely to get rid of this `unreachable!()`?

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 6 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:2058]: It is now equal because the reject response is never in the stream because it is generated locally from the reject signal and then inducted immediately.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 5 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: Done

File: rs/messaging/src/routing/stream_handler.rs (hunk 2 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: Had to change it again because of supporting the old and the new way simultaneously. The return value is still optional though and I added a doc comment explaining this choice.

File: rs/messaging/src/routing/stream_handler.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: Now moot with the old way back in.

File: rs/messaging/src/routing/stream_handler.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (alin-at-dfinity) [reply rs/messaging/src/routing/stream_handler.rs]: Exactly. I would introduce an explicit `MIN_SUPPORTED_CERTIFICATION_VERSION` for that, as `all_supported_versions()` is mostly intended for exhaustive tests. In production code you would simply do `min <= version <= max`.

[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: The old way with reject responses is back in for `ReplicatedState` of `CertificationVersion::V18` and below. Including all of the old versions of tests. I didn't change that much of the code in tests but in multiple places and I don't much like the idea of multiple case distinctions per test. So I just put the old tests in there, put a `legacy_` affix on them and a `TODO` in top. So when we eventually want to get rid of it you can just remove whatever has this `TODO` on it and then the compiler will tell what else is not used anymore.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 4 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs]: As I see it remodelling what `available_guaranteed_response_memory()` does here is out of place since there should be a test for that somewhere else. The new way more explicitly shows what the difference is in the context of what we are testing here without overloading this test with information about an otherwise unrelated function.

File: rs/messaging/src/routing/stream_handler/tests.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler/tests.rs:1663]: The old versions are also back in with a `legacy_` affix testing for `CertificationVersion::V18` and below.

File: rs/messaging/src/routing/stream_handler.rs [hunk 10]
```diff
@@ -610,14 +610,14 @@ impl StreamHandlerImpl {
...
                     ) {
-                        Ok(()) => {
+                        None => {
                             // Reject response successfully inducted or dropped.
                         }
-                        Err((StateError::CanisterMigrating { .. }, reject_response)) => {
+                        Some((StateError::CanisterMigrating { .. }, reject_response)) => {
                             // Canister is being migrated, reroute reject response.
                             reroute_response(reject_response, state, streams, &self.log);
                         }
```
Comment (derlerd-dfinity) [inline rs/messaging/src/routing/stream_handler.rs:619]: Is it really possible that a response that is already rerouted needs to be rerouted again?

Comment (derlerd-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:619]: Strike that - I read that as being part of the inline function `reroute_response` above and thought we'd end up with a recursive call.

File: rs/messaging/tests/memory_tests.rs [hunk 11]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+/// signals for requests in the reverse stream to the local_subnet. But since we migrated the
+/// `migrating_canister` to the remote subnet, the locally generated reject responses fail to
+/// induct and are rerouted into the stream to the remote subnet. The remote subnet eventually picks
+/// them up and inducts them into `migrating_canister` leaving no hanging calls after some more rounds.
+///
+/// If there are hanging calls after a threshold number of rounds, there is most likely a bug
+/// connected to reject signals for requests, specifically with the corresponding exceptions due to
+/// canister migration.
+#[test]
+fn check_calls_conclude_with_migrating_canister() {
```
Comment (stiegerc) [inline rs/messaging/tests/memory_tests.rs:198]: One general comment on the files in 'rs/messaging/tests/'. The names are arbitrary and perhaps this should be refactored into just one file 'rs/messaging/tests/messaging.rs' since this test isn't a 'memory test', it's also not exactly clear what a 'queue test' is supposed to be and why is it in message routing when queues are in replicated state and so on.

What I made here is such that what I want to check (the rerouting of locally generated reject responses into the stream to the subnet where the canister moved to) but also such that I can just use the framework I already have. In principle there is nothing wrong with it.

But from what we were discussing and what might also be interesting to see in action is a great deal of such reject responses ending up in one stream by making one canister send messages to lots canister on lots of subnets, and then checking whether all of the corresponding reject responses land in the same stream. I mean this test does check that but with just one subnet. I didn't go for this because I might have to extend the framework quite a bit and also it would make the test run for a long time just to generate so many messages so I figured it's not worth it.

File: rs/messaging/src/routing/stream_handler.rs [hunk 12]
```diff
@@ -766,9 +763,27 @@ impl StreamHandlerImpl {
...
+                    };
+                    stream.push_reject_signal(reason);
+                }
+                Some((StateError::CanisterMigrating { .. }, RequestOrResponse::Response(_))) => {
+                    // Unable to deliver a response due to migrating canister, push reject signal.
+                    stream.push_reject_signal(RejectReason::CanisterMigrating);
                 }
-                Err((_, RequestOrResponse::Response(_))) => {
+                Some((_, RequestOrResponse::Response(_))) => {
                     unreachable!("No signals are generated for response induction failures except for CanisterMigrating");
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler.rs:787]: I would handle both unreachable cases (this and the one above) consistently: either both `unreachable!()` or both returning `RejectReason::Unknown`.

File: rs/messaging/src/routing/stream_handler.rs [hunk 13]
```diff
@@ -792,13 +807,20 @@ impl StreamHandlerImpl {
         }
     }
 
+    /// Inducts a message into `state`. This may fail and log an error for cases where there is
+    /// nothing we can do about it; for other cases it may return `(StateError, RequestOrResponse)`
+    /// as it is returned by `ReplicatedState::push_input()` upon failure such that we may generate
+    /// a reject response or push a reject signal.
```
Comment (alin-at-dfinity) [inline rs/messaging/src/routing/stream_handler.rs:813]: I would try to be clearer about the 3 possible outcomes:
1. Message successfully inducted, returns `None`.
2. Message silently dropped (e.g. late best-effort response), returns `None`.
3. Failed to induct message (error or canister migrating), returns `(StateError, RequestOrResponse)`. Caller is expected to produce a reject response or a reject signal.

File: rs/messaging/tests/memory_tests.rs [hunk 14]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+        return fixture.failed_with_reason("found hanging calls in the records");
     }
 
     // After the fact, all memory is freed and back to 0.
     fixture.expect_guaranteed_response_message_memory_taken_at_most("Final check", 0, 0)
 }
 
+/// Runs a state machine test with two subnets, a local subnet with 2 canisters installed and a
+/// remote subnet with 5 canisters installed. All canisters, except one local canister referred to
+/// as `migrating_canister` is stopped.
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:181]:
```suggestion
/// remote subnet with 5 canisters installed. All canisters, except one local canister referred to
/// as `migrating_canister` are stopped.
```

File: rs/messaging/tests/memory_tests.rs [hunk 15]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+///
+/// In the first phase a number of rounds are executed on both subnets, including XNet traffic with
+/// the `migrating_canister` making random calls to all installed canisters (since all calls are
+/// rejected except those to self, downstream calls are disabled).
+///
+/// For the second phase, `migrating_canister` stops making calls and is then migrated to the
+/// remote subnet. Since all other canisters are stopped, there are bound to be a number of reject
+/// signals for requests in the reverse stream to the local_subnet. But since we migrated the
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:189]: Nit: We use "reverse stream" when we're dealing with pairs of streams (and one is obviously the "[outbound] stream" and the other the "reverse stream"). In this case, you should probably use "stream to the local subnet", "incoming stream(s)" or "inbound stream(s)".

File: rs/messaging/tests/memory_tests.rs [hunk 16]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+        fixture.tick();
+    }
+
+    // Stop making calls and migrate `migrating_canister`.
+    fixture
+        .set_max_calls_per_heartbeat(migrating_canister, 0)
+        .unwrap();
+    fixture.migrate_canister(migrating_canister);
+
+    // Tick until all calls or concluded.
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:245]:
```suggestion
    // Tick until all calls have concluded.
```

File: rs/messaging/tests/memory_tests.rs [hunk 17]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+    for counter in 0.. {
+        fixture.tick();
+        if fixture.open_call_contexts_count().values().sum::<usize>() == 0 {
+            break;
+        }
+        assert!(counter < SHUTDOWN_PHASE_MAX_ROUNDS);
+    }
+
+    // Check the records agree on 'no hanging calls'.
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:254]:
```suggestion
    // Check that the records agree on 'no hanging calls'.
```

File: rs/messaging/tests/memory_tests.rs [hunk 18]
```diff
@@ -357,35 +460,43 @@ impl Fixture {
...
     }
 
-    /// Collects the metrics for all installed canisters on the fixture.
-    pub fn collect_metrics(&self) -> BTreeMap<CanisterId, CanisterMetrics> {
-        self.canisters()
-            .into_iter()
-            .map(|canister| (canister, extract_metrics(&self.query_records(canister))))
-            .collect()
+    /// Force Queries the records from `canister` by attempting to query them; if it fails start
+    /// the canister and query them again.
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:489]:
```suggestion
    /// Force queries the records from `canister` by first attempting to query them; if it fails, start
    /// the canister and try querying them again.
```

File: rs/messaging/tests/memory_tests.rs [hunk 19]
```diff
@@ -469,6 +580,46 @@ impl Fixture {
         }
     }
 
+    /// Migrates `canister` from the `local_env` (`remote_env)  to `remote_env` (`local_env`).
```
Comment (alin-at-dfinity) [inline rs/messaging/tests/memory_tests.rs:583]:
```suggestion
    /// Migrates `canister` between `local_env` and `remote_env` (either direction).
```

**Review by alin-at-dfinity — APPROVED** [review]
LGTM, just a bunch of nitpicks.

File: rs/messaging/tests/memory_tests.rs (hunk 14 above)
Comment (stiegerc) [reply rs/messaging/tests/memory_tests.rs]: Also put a comma there.

File: rs/messaging/src/routing/stream_handler.rs (hunk 12 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:788]: We don't want to panic for requests because the error is just passed along from `ReplicatedState::push_input()`; for responses the code in `induct_message_impl()` is such that only `StateError::CanisterMigrating { .. }` can come up so it would have to be a local change that triggers this. You wouldn't get away with it without changing unit tests which is why I felt it's ok to panic and we are not supposed to generate reject signals for responses for anything other than canister migrating.

OTOH we have to handle all combinations of response and `RejectSignal` anyway so we could also just treat all messages the same and remove the special cases for responses. If anything other than canister migrating gets through (it's not possible from the code as it is now though), `handle_rejected_messages()` on the other subnet will log an error so it won't go unnoticed.

I'm fine with either since the panic can't trigger on mainnet since it won't pass the CI if there is an issue. Then again perhaps we shouldn't rely on it if we don't have to. WDYT @derlerd-dfinity ?

File: rs/messaging/src/routing/stream_handler.rs [hunk 20]
```diff
@@ -766,9 +763,28 @@ impl StreamHandlerImpl {
...
+                Some((err, RequestOrResponse::Request(_))) => {
+                    // Unable to induct a request, push a reject signal.
+                    let reason = match err {
+                        StateError::CanisterMigrating { .. } => RejectReason::CanisterMigrating,
+                        StateError::CanisterNotFound(_) => RejectReason::CanisterNotFound,
+                        StateError::CanisterStopped(_) => RejectReason::CanisterStopped,
+                        StateError::CanisterStopping(_) => RejectReason::CanisterStopping,
+                        StateError::QueueFull { .. } => RejectReason::QueueFull,
+                        StateError::OutOfMemory { .. } => RejectReason::OutOfMemory,
+                        // Unreachable.
```
Comment (oggy-dfin) [inline rs/messaging/src/routing/stream_handler.rs:777]: I know I'm late to the party, and maybe I'm repeating a discussion from two months ago (sorry if I'm doing that), but: why not have `induct_impl` return a `RejectReason` and then do the conversion, getting rid of the unreachable stuff?

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:777]: That's how I had it before I realized we have to support certification version 18 and below too, i.e. generate reject responses rather than signals. This is implemented by turning `StateError` into a string which is why I went with  returning an the error, not a signal.

OTOH you could perhaps change that to generate a reject response from the signal rather than the error but the error contains more information than a signal so I can't produce the exact same message. I'll give that a try; if it doesn't significantly bloat the diff or causes problems in execution or something this might be the cleaner solution.

File: rs/messaging/src/routing/stream_handler.rs (hunk 12 above)
Comment (derlerd-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:788]: Our policy is not to crash if there is a meaningful way to go ahead. So if returning `RejectReason::Unknown` is an option that works this is the way to go. 

Lets maybe discuss this in more detail. Perhaps also @oggy-dfin has an opinion?

File: rs/messaging/tests/memory_tests.rs [hunk 21]
```diff
@@ -146,15 +162,102 @@ fn check_guaranteed_response_message_memory_limits_are_respected_impl(
...
+///
+/// In the first phase a number of rounds are executed on both subnets, including XNet traffic with
+/// the `migrating_canister` making random calls to all installed canisters (since all calls are
+/// rejected except those to self, downstream calls are disabled).
+///
+/// For the second phase, `migrating_canister` stops making calls and is then migrated to the
+/// remote subnet. Since all other canisters are stopped, there are bound to be a number of reject
+/// signals for requests in the stream to the local_subnet. But since we migrated thec`migrating_canister`
+/// to the remote subnet, the locally generated reject responses fail tocinduct and are rerouted into the
+/// stream to the remote subnet. The remote subnet eventually pickscthem up and inducts them into
```
[the remark this answers is not in the document]
Comment (stiegerc) [reply rs/messaging/tests/memory_tests.rs]: No idea ¯\_(ツ)_/¯

**Review by berestovskyy — APPROVED** [review]

File: rs/messaging/src/routing/stream_handler.rs (hunk 20 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: So when I initially put support for older certification versions back in, I made it exactly the way it was before. But that is actually not necessary; I'm not sure why I did it like that TBH.

I changed it now such that `induct_message_impl()` returns `Option<(RejectReason, RequestOrResponse)>` and thereby moved the conversion from `StateError` to `RejectReason` in there directly (as it was at an earlier stage). So what happens
- for certification version 19 and after is that it just uses this `RejectReason` to push a `RejectSignal`. On the other subnet it then eventually generates and inducts a reject response locally when gc'ing the corresponding request using the reject signal to identify the proper pair of `reason` and ` request` using `generate_reject_response_for(reason, request)`.
- for certification version 18 and below it uses the `reason` and `request` in place in `induct_message()` on this subnet to generate a reject response using `generate_reject_response_for(reason, request)` and then pushes that into the stream to the other subnet.

So from the point of view of a canister on the other subnet I get exactly the same reject response for all certification versions. However note that this PR slightly changes the message contained in the reject response for all certifications versions (before I changed it only for version 19 at up). But this was discussed and is not a problem as far as I know.

For example for an 'out of memory' reject response, we previously included how much memory was left on the subnet because this information is available in `StateError::OutOfMemory`. But since a reject signal is just a stream index and a reject reason, this information is lost and thus not included anymore (note: it still appears in the logs). However it was quite useless anyway because a request always consumes 2MB due to the reservation so it's not like a canister could have constructed a smaller request to fit it in (and also the situation would have changed in the mean time anyway).

This change allowed me to remove `StateError::CanisterMigrating` (which I added some time ago) since the only place it was used is to return it from `induct_message_impl()` and it otherwise caused only some minor changes in old tests to adapt for the new way a reject response is generated. I also slightly changed the message used for logging since before it claimed to produce a reject response which is actually false depending on the certification version. It is now neutral such that it both fits generating a reject response and pushing a reject signal.

I think this is a good change and I'm fairly certain @alin-at-dfinity would agree.

File: rs/messaging/src/routing/stream_handler.rs (hunk 12 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:764]: Note that there is no code path into this `unreachable!()`. It is literally unreachable. The only combination of `(RejectReason, RequestOrResponse::Response)` `induct_message_impl()` returns is `(RejectReason::CanisterMigrating, RequestOrResponse::Response)` since for all other combinations the message is silently dropped. It is more of a documentation of this behavior I suppose than an actually functionality.

Comment (derlerd-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:764]: Yeah, I saw it as well. I guess this might be an argument for actually not treating both cases the same. Here unreachable might arguably be fine because the behaviour we rely on is encapsulated within `stream_handler.rs`. For the other case, i.e., the one above where we return `RejectReason::Unknown` we'd have to rely that nobody changes `push_input` in a way that we don't expect.

Comment (oggy-dfin) [reply rs/messaging/src/routing/stream_handler.rs:764]: Given that it's all in a single `impl` I think it's passable. I mean we could also type `induct_message_impl` more precisely (e.g., return a sum type `(Response * CanisterMigrating) + (Request * (CanisterMigrating + CanisterNotFound + ...))`) but maybe an overkill here.

Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs:764]: AFAICT you can't do that in Rust as of now. Maybe I'm not up to date here?

I read a while ago that it is an often requested feature to have individual variants as their own type. Then perhaps you could return a kind of match statement. Something like
```Rust
fn my_fn() -> (_, RequestOrResponse::Request(_)) | (RejectReason::CanisterMigrating, RequestOrResponse::Response(_))`.
```
And then the compiler could use that in the `match` statement. For now, it seems we are stuck with just setting impossible combinations to `unreachable!()`.

File: rs/messaging/src/routing/stream_handler.rs (hunk 20 above)
Comment (stiegerc) [reply rs/messaging/src/routing/stream_handler.rs]: One additional note about the return type here. It was a bit weird to return `Result<(), (StateError, RequestOrResponse)>` because returning `Ok(())` can mean _successfully inducted_ or _silently_dropped_ so essentially it means _no_further_action_required_. Then I changed it to `Option<(StateError, RequestOrResponse)>` since that seems more idiomatic for this case, but then it looks a bit weird that we return an error as optional.

So long story short `Option<(RejectReason, RequestOrResponse)>` removes all the weirdness since it's not an error anymore, but it still includes the implication that the caller is expected to do something with it unless it is `None`.

File: rs/messaging/src/routing/stream_handler.rs (hunk 12 above)
Comment (oggy-dfin) [reply rs/messaging/src/routing/stream_handler.rs:764]: Yeah I was using just some made up syntax that'd need to become a bit more heavy weight in Rust. Concretely, you'd need `induct_message_impl` to return `Option<Either<(Response, ResponseErrors), (Request, RejectReason)>>` (where `Either` is basically the same thing as `Result`, just named differently since it's not supposed to distinguish between errors, just give you a sum). But reading that return type is a bit painful, and you'd also need to have an `enum ResponseErrors { CanisterMigrating }`, which is what I mean by it being heavy. 

But then you could match on the result of `induct_message_impl` without having any unreachable branches.

File: rs/messaging/src/routing/stream_handler.rs [hunk 22]
```diff
@@ -615,14 +615,14 @@ impl StreamHandlerImpl {
                         state,
                         available_guaranteed_response_memory,
                     ) {
-                        Ok(()) => {
+                        None => {
```
Comment (derlerd-dfinity) [inline rs/messaging/src/routing/stream_handler.rs:618]: I just thought about this case here once more. If I'm understanding correctly we might end up dropping responses here that we shouldn't drop once we enable reject signals for requests. 

IIUC `available_guaranteed_response_memory` is not updated when we garbage collect a message from a stream during induction here for simplicity (see the comment on the computation of `available_guaranteed_response_memory`). This means that the memory for a rejected request that we garbage collected before entering this function is not yet added to the available memory yet at this point, which, in turn, seems to suggest that we might end up in situations where we can't push the request here because we're seemingly out of memory.

Am I missing something?

Comment (derlerd-dfinity) [reply rs/messaging/src/routing/stream_handler.rs:618]: I think what I have missed is that `push_input` will ignore available memory for responses given that a response will always free memory.
